mirrord/intproxy/src/failover_strategy.rs

```diff
@@ -0,0 +1,165 @@
+                    Err(ProxyStartupError::ConnectionAcceptTimeout)?;
+                },
+                _ = time::sleep(idle_timeout), if failover.any_connection_accepted && !failover.has_layer_connections() => {
+                    tracing::info!("Reached the idle timeout with no active layer connections");
+                    break;
+                },
+            }
+        }
+
+        std::mem::drop(failover.layer_initializer);
+        std::mem::drop(failover.layers);
+
+        tracing::info!("Collecting background task results before exiting");
+        let results = failover.background_tasks.results().await;
+
+        for (task_id, result) in results {
+            tracing::trace!(
+                %task_id,
+                ?result,
+                "Collected a background task result",
+            );
+        }
+
+        Ok(())
+    }
+
+    async fn handle_task_update(
+        &mut self,
+        task_id: MainTaskId,
+        update: TaskUpdate<ProxyMessage, ProxyRuntimeError>,
+    ) {
+        match (task_id, update) {
+            (MainTaskId::LayerConnection(LayerId(id)), TaskUpdate::Finished(Ok(()))) => {
+                tracing::trace!(layer_id = id, "Layer connection closed");
+                self.layers.remove(&LayerId(id));
+            }
+            (task_id, TaskUpdate::Finished(res)) => match res {
+                Ok(()) => {
+                    tracing::error!(%task_id, "One of the main tasks finished unexpectedly");
+                }
+                Err(TaskError::Error(error)) => {
+                    tracing::error!(%task_id, %error, "One of the main tasks failed");
+                }
+                Err(TaskError::Panic) => {
+                    tracing::error!(%task_id, "One of the main tasks panicked");
+                }
+            },
+
+            (_, TaskUpdate::Message(msg)) => self.handle(msg).await,
+        }
+    }
+
+    async fn handle(&mut self, msg: ProxyMessage) {
+        match msg {
+            ProxyMessage::NewLayer(new_layer) => {
+                self.any_connection_accepted = true;
+                let tx = self.background_tasks.register(
+                    LayerConnection::new(new_layer.stream, new_layer.id),
+                    MainTaskId::LayerConnection(new_layer.id),
+                    IntProxy::CHANNEL_SIZE,
+                );
+                self.layers.insert(new_layer.id, tx);
+            }
+            ProxyMessage::FromLayer(FromLayer {
+                message: msg @ LayerToProxyMessage::Incoming(_),
+                ..
+            }) => {
+                tracing::info!(message = ?msg, "Proxy in failover mode, ignoring a message");
+            }
+            ProxyMessage::FromLayer(FromLayer {
+                message_id,
+                layer_id,
+                ..
+            }) => {
+                self.update_layer_on_error(layer_id, message_id).await;
+            }
```

Comment:
```suggestion
            ProxyMessage::FromLayer(message) => {
                self.update_layer_on_error(message).await;
            }
```
This needs more a granular check of the message kind, I think best done in `update_layer_on_error`